### DISCUSSION TITLE
Custom icons for sandboxed app shortcuts - thoughts?

### DISCUSSION THREAD

**Discussion by e-t-l** [sandboxie-plus/Sandboxie#3379] [Various ideas and suggestions]
I am thinking about extending SBIE+'s ability to generate shortcuts to sandboxed apps, to give said shortcuts a unique icon that shows what app is being sandboxed (instead of the default SBIE+ icon, which is the current behavior). 

For example, this is the current Desktop icon for the generated "sandboxed browser" shortcut, which targets Firefox:
![sbie](https://github.com/sandboxie-plus/Sandboxie/assets/40775958/b924d00d-788c-43a4-86c9-cefce3679653)

It's not readily apparent what app will be launched by this shortcut!

I have two ideas (actually I have more, but these are my top two) for how the shortcut would look instead. The images below are mock-ups, but the idea would be that _anytime_ the user creates a shortcut to a sandboxed file or app in Sandboxie-Plus, it would generate a Sandbox Shortcut ICO based off the ICO of the target.

### Idea no.1: The "shortcut arrow" overlay
A sized-down, relatively inconspicuous SBIE+ icon overlaying the bottom right corner, similar to how Windows puts an arrow symbol on the bottom left of .lnk icons. (Or the overlay could replace the shortcut arrow in the bottom left! Open for discussion.)

![ffbox2](https://github.com/sandboxie-plus/Sandboxie/assets/40775958/3e7cf9e9-85f4-4fb1-900a-493d7b663868)

### Idea no. 2: ICO-in-the-box
The target icon (slightly sized-down) is depicted as sitting inside the "sandbox" part of the SBIE+ icon.

![ffbox1](https://github.com/sandboxie-plus/Sandboxie/assets/40775958/21f3faec-1159-4a4e-98a2-7e8ae7d20ec2)

.
.
.

@DavidXanatos is this something you'd be interested in having? (Also, would this kind of contribution be worthy of a Contributor License?)

**Comment by e-t-l** [comment]
Forgot to mention this, but if there's no clear preference for one version above all others, one possibility is to add an option to the SBIE+ Settings to let the user select whether shortcuts should use the Default option, Variant no.1, or Variant no.2.

**Comment by DavidXanatos** [comment]
I'm not sure how well this will look on non retina displays, a standard icon is 32x32 so we would need a at most 16x16 better 12x12 overlay.
If its made well and creates nice looking icons that would be cool and as that will surely require some significant afford to make it look good its worth a Contributor License.

Sandman would need to generate *.ico files with different resolutions, 16x16, 32x32, 48x48, 64x64 and may be bigger save them in a  cache location and make the *.lnk files use them. 
For the 16x16 icon I would suggest to just put a 1 pixel square yellow border around the original icon and done, as I did that for sandboxed tray icons.

**Comment by e-t-l** [comment]
> 16x16 icon I would suggest to just put a 1 pixel square yellow border around the original icon and done

That's a good idea

> save them in a cache location and make the *.lnk files use them

Yeah, I was thinking it would just be a folder in the SBIE install directory that holds all the ICO files, with all the shortcuts pointing back to that location. That way it would avoid regenerating an icon that has already been generated, and all the icons could be changed/updated/deleted as a group if need be.

Before I get started, I have a few questions:
1. Do you have a preference for which style to go with? (Considering low-resolution scenarios, I think Idea no.2 would be better)
2. This is a pretty big repo, and I won't pretend to understand how everything is organized. Is there anywhere specific that you'd like me to start putting this icon-generating code? 
3. What would you say is the best way to understand the differences between [`CBoxObject::MakeShortcut`](https://github.com/sandboxie-plus/Sandboxie/blob/10b2cfb7673b670df21f924336080c3c457a36e6/SandboxiePlus/SandMan/Engine/BoxObject.cpp#L77 "/SandboxiePlus/SandMan/Engine/BoxObject.cpp#L77"), [`CSbieUtils::CreateShortcut`](https://github.com/sandboxie-plus/Sandboxie/blob/10b2cfb7673b670df21f924336080c3c457a36e6/SandboxiePlus/QSbieAPI/SbieUtils.cpp#L493 "/SandboxiePlus/QSbieAPI/SbieUtils.cpp#L493"), [`CSbieView::CreateShortcutEx`](https://github.com/sandboxie-plus/Sandboxie/blob/10b2cfb7673b670df21f924336080c3c457a36e6/SandboxiePlus/SandMan/Views/SbieView.cpp#L1580 "/SandboxiePlus/SandMan/Views/SbieView.cpp#L1580"), [`CFileListCtrl::OnCmdCreateShortcut`](https://github.com/sandboxie-plus/Sandboxie/blob/10b2cfb7673b670df21f924336080c3c457a36e6/Sandboxie/apps/control/FileListCtrl.cpp#L1408 "/Sandboxie/apps/control/FileListCtrl.cpp#L1408"), [`CShellDialog::CreateShortcut`](https://github.com/sandboxie-plus/Sandboxie/blob/480d416234ea1202fdefd1820c313e7ba3edeef4/Sandboxie/apps/control/ShellDialog.cpp#L1195 "/Sandboxie/apps/control/ShellDialog.cpp#L1195"), and [`CShellDialog::CreateDesktopShortcut`](https://github.com/sandboxie-plus/Sandboxie/blob/480d416234ea1202fdefd1820c313e7ba3edeef4/Sandboxie/apps/control/ShellDialog.cpp#L1066 "/Sandboxie/apps/control/ShellDialog.cpp#L1066")? (particularly these last two in ShellDialog.cpp, which seem to call *each other*, at [#L1101](https://github.com/sandboxie-plus/Sandboxie/blob/480d416234ea1202fdefd1820c313e7ba3edeef4/Sandboxie/apps/control/ShellDialog.cpp#L1101 "CreateShortcut(path, LinkName, BoxName, LinkPath, IconPath, IconIndex, WorkDir);") and [#L1201](https://github.com/sandboxie-plus/Sandboxie/blob/480d416234ea1202fdefd1820c313e7ba3edeef4/Sandboxie/apps/control/ShellDialog.cpp#L1201 "CreateDesktopShortcut(BoxName, LinkPath, IconPath, IconIndex, WorkDir);"), respectively)
4. Do any of the above *not* need custom icon integration?
5. Are there any other functions that deal with shortcut creation that I haven't listed, where a custom icon would be relevant? (I know there's [`CSbieUtils::GetStartMenuShortcut`](https://github.com/sandboxie-plus/Sandboxie/blob/480d416234ea1202fdefd1820c313e7ba3edeef4/SandboxiePlus/QSbieAPI/SbieUtils.cpp#L547 "/SandboxiePlus/QSbieAPI/SbieUtils.cpp#L547") but I think that one wouldn't need custom icons, right?)
